### PR TITLE
feat: Integrate XHGui into DDEV, fixes #6894

### LINKED ISSUE #6894: XHGui support integrated in DDEV core
Related: 
- #3782 
- [TYPO3 Community Budget Process Announcement](https://typo3.org/article/four-ideas-to-be-funded-in-quarter-1-2025)

The TYPO3 Community Ideas Process has selected DDEV XHGui  integration as a project for Q1 2025. 

This issue will serve as the project plan for this.

Goals:
- [x] Improve the existing [ddev-xhgui](https://github.com/ddev/ddev-xhgui) add-on with a number of features, including removing the requirement for additional configuration.
- [x] Add xhgui command to DDEV. It should create required database if it doesn't exist, and enable xhprof. 
- [ ] Deprecate/hide the xhprof command
- [x] Avoid breaking the workflow of existing users that depend on xhprof in the traditional technique
- [x] Update docs for performance
- [ ] Deprecate the ddev-xhgui add-on
- [ ] Run xhgui container with the current user (instead of root)
- [x] Add bash to the xhgui built image
- [x] Consider pegging xhgui image instead of using `latest`.
- [x] Consider installing xhgui in the web container instead of using separate container.
- [x] Xhgui should not be in the prod/hardened image

Notes:
- [x] [xhgui/xhgui](https://hub.docker.com/r/xhgui/xhgui) is alpine-based and hasn't been pushed in a year, and has no information about its build. The [org](https://hub.docker.com/u/xhgui) also has no information. 
- [x] The [github repo](https://github.com/perftools/xhgui) appears to be maintained by the same user ([glensc](https://github.com/glensc)) who pushes the xhgui image.
- [x] His php-profiler shows recent but sporadic activity, https://github.com/perftools/php-profiler
- [x] Wow, xhgui has been around for a long time!

### PR DESCRIPTION
**Pull request by rfay**
## The Issue

- #6894 

Honors to @tyler36, who pioneered and then matured the xhgui add-on until it could be brought into DDEV here.

## How This PR Solves The Issue

Add integration for xhgui

## TODO

- [x] Handle case of existing ddev-xhgui add-on
- [x] Tests
- [x] Fix `ddev xhgui -j`, perhaps add some support
- [x] Make the xhgui port configurable
- [x] Make work without router like mailpit
- [x] FIgure out if `ddev xhgui` should launch browser. Adjust as needed
- [ ] Report to TYPO3 folks (for typo3.org) about the outcome at [link](https://typo3.org/community/teams/communication-coordination/content-request), [using](https://typo3.org/article/community-budget-idea-report-image-rendering-improvements) [these](https://typo3.org/article/community-budget-report-future-of-the-staticfilecache-extension) [examples](https://typo3.org/project/news/budget-idea).

## Manual Testing Instructions

Docs suggestions:
*  https://ddev--7069.org.readthedocs.build/en/7069/users/debugging-profiling/xhprof-profiling/#simplest-xhprof-usage-with-xhgui
*  https://ddev--7069.org.readthedocs.build/en/7069/users/usage/commands/#xhgui
* https://ddev--7069.org.readthedocs.build/en/7069/users/configuration/config/#xhprof_mode

- [ ] `ddev config global --xhprof-mode=xhgui`
- [ ] `ddev xhgui on`
- [ ] `ddev xhgui status`
- [ ] `ddev xhgui launch`
- [ ] `ddev xhprof on`
- [ ] Validate that xhprof still works with xhprof-mode=prepend
- [ ] Install ddev/ddev-xhgui and then `ddev start` to see removal
- [x] Test with a Postgres project

## Automated Testing Overview

- [x] TestDdevXhprofXhguiEnabled
- [x] TestCmdXHGui
- [x] TestXHGuiSetup

## Release/Deployment Notes

- [ ] Note that the ddev-xhgui add-on will not be compatible.
- [ ] Announce to supporting TYPO3 community!
- [ ] Blog and screencast

### CODE REVIEW

**Comment by rfay** [comment]
Moved to "Ready for Review" even though certainly not perfect. Folks who have used xhgui will be great reviewers. Hope to get @tyler36 's expertise in here.

**Comment by tyler36** [comment]
Did some basic testing with DDEV `v1.24.3-54-g485c3ad89`

- ✅ `ddev xhgui` asks to add global configuration
- ✅ `ddev config global --xhprof-mode=xhgui` updates config
- ✅ `ddev xhgui on` starts debugging, confirmed with `ddev xhgui launch`
- ✅ `ddev xhgui status` correctly reports on
- ✅ `ddev xhgui off` and am no longer able to access website via `ddev xhgui lanuch`
- ✅  delete project database, changed to Postgres 16 and correctly ran `ddev xhgui launch`
- ✅ `ddev xhgui -h` and `ddev xhgui --help` both correctly display help messgae
- ❓ `ddev xhgui -j` launches xhgui website.

**Comment by rfay** [comment]
Thanks @tyler36 - I didn't implement `ddev xhgui -j`, what would you expect it to do?

`ddev xhgui` and `ddev xhgui launch` do kind of the same as `ddev xdebug` and `ddev xdebug on`

**Comment by tyler36** [comment]
> Thanks @tyler36 - I didn't implement ddev xhgui -j, what would you expect it to do?

Not sure what I expected. Just felt a little strange that it opened my browser.

File: pkg/ddevapp/schema.json [hunk 1]
```diff
@@ -411,7 +411,14 @@
...
     },
-    "web_environment": {
+    "xhprof_mode": {
+      "description": "Defines the xhprof usage mode used by the project.",
+      "type": "string",
+      "enum": [
+        "prepend",
+        "xhgui"
+      ]
+    },    "web_environment": {
```
Comment (stasadev) [inline pkg/ddevapp/schema.json:421]: wrong formatting.

**Review by stasadev — CHANGES_REQUESTED** [review]
Wow!

This change also needs an update to the global `schema.json`

(I will test edge cases later today.)

**Comment by rfay** [comment]
@tyler36 I think I'll change `ddev xhgui` to enable and show status and not launch browser. I guess the reason it did that was the command in the add-on did that (but that's all it did).

**Comment by stasadev** [comment]
> I think I'll change ddev xhgui to enable and show status and not launch browser.

I actually like that it opens the browser. No need to remember anything. (See my suggestion for `-j`).

File: pkg/config/types/xhprof_mode.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+		}
+	}
+
+	return false
+}
+
+// CheckValidXHProfMode checks to see if the given xhprof mode option
+// is valid and returns an error in case the value is not valid.
+func CheckValidXHProfMode(xhprofMode string, configType ConfigType) error {
+	// TODO
```
Comment (stasadev) [inline pkg/config/types/xhprof_mode.go:56]:
```suggestion
```

File: pkg/config/types/xhprof_mode.go [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+// IsValidXHProfMode checks to see if the given xhprof mode
+// option is valid.
+func IsValidXHProfMode(XHProfMode string, configType ConfigType) bool {
+	if XHProfMode == XHProfModeEmpty {
+		return true
+	}
+
+	for _, o := range ValidXHProfModeOptions(configType) {
+		if XHProfMode == o {
```
Comment (stasadev) [inline pkg/config/types/xhprof_mode.go:45]: For readability:

```suggestion
	for _, option := range ValidXHProfModeOptions(configType) {
		if XHProfMode == option {
```

File: docs/content/users/configuration/config.md [hunk 4]
```diff
@@ -735,3 +745,27 @@ Examples:
...
+| -- |---------| --
+| :octicons-file-directory-16: project<br>:octicons-globe-16: global | `8142`  | Can be changed to avoid a port conflict.
+
+## `xhprof_mode`
+
+Whether XHProf should use `prepend` or `xhgui` mode.
+
+| Type | Default | Usage
+| -- | -- | --
+| :octicons-file-directory-16: project<br>:octicons-globe-16: global | `false` | `prepend` or `xhgui`.
```
Comment (stasadev) [inline docs/content/users/configuration/config.md:771]:
```suggestion
| :octicons-file-directory-16: project<br>:octicons-globe-16: global | `prepend` | `prepend` or `xhgui`.
```

File: .goreleaser.yml [hunk 5]
```diff
@@ -482,6 +482,16 @@ dockerhub:
...
 
+  #ddev-xhgui
+  - images:
+      - '{{ .Env.DOCKER_ORG }}/ddev-xhgui'
+    username: '{{ .Env.DOCKERHUB_USERNAME }}'
+    secret_name: "DOCKERHUB_TOKEN"
+    description: xhgui facility for DDEV integration
+    full_description:
+      from_url:
+        url: https://github.com/ddev/ddev-xhgui/raw/main/README.md
```
Comment (stasadev) [inline .goreleaser.yml:493]:
```suggestion
      from_file:
        path: ./containers/ddev-xhgui/README.md
```

File: cmd/ddev/cmd/xhgui.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/ddev/ddev/pkg/util"
+	"github.com/spf13/cobra"
+)
+
+// DdevXHGuiCmd represents the xhgui command.
+var DdevXHGuiCmd = &cobra.Command{
+	ValidArgsFunction: ddevapp.GetProjectNamesFunc("active", 1),
+	Use:               "xhgui [on|off|launch|status]",
+	Short:             "Starts or checks status of XHGui performance monitoring and UI",
+	//Long:              `Starts a shell session in the container for a service. Uses web service by default. To start a shell session for another service, run "ddev ssh --service <service>`,
```
Comment (stasadev) [inline cmd/ddev/cmd/xhgui.go:18]:
```suggestion
```

File: containers/ddev-webserver/ddev-webserver-base-files/usr/local/bin/enable_xhprof [hunk 7]
```diff
@@ -4,6 +4,13 @@ phpdismod blackfire xdebug
 mkdir -p ${XHPROF_OUTPUT_DIR}
 phpenmod xhprof
 killall -USR2 php-fpm 2>/dev/null || true
-echo "Enabled xhprof.
-After each web request or CLI process you can see all runs, most recent first, at
+echo "Enabled xhprof with xhprof_mode=${DDEV_XHPROF_MODE}"
+case "${DDEV_XHPROF_MODE}" in
+  xhgui)
+    echo "Use 'ddev xhgui' to launch browser for xhprof results."
```
Comment (stasadev) [inline containers/ddev-webserver/ddev-webserver-base-files/usr/local/bin/enable_xhprof:10]: The image needs a new push, I see the old message `to lanch browser`

**Review by stasadev — CHANGES_REQUESTED** [review]
Old approach doesn't work for me.
(I don't see any records when I refresh the page.)

```
ddev config global --xhprof-mode-reset && ddev restart && ddev xhprof && ddev launch /xhprof
```

---

And noticed some inconsistency in `--xhprof-mode-reset` (it may also exist in another `-reset` flag):

```
ddev config global --xhprof-mode=xhgui
ddev config --xhprof-mode=prepend
# doesn't reset .ddev/config.yaml
ddev config --xhprof-mode-reset
# but this works
ddev config --xhprof-mode=""
```

File: containers/ddev-webserver/ddev-webserver-base-files/usr/local/bin/enable_xhprof (hunk 7 above)
Comment (rfay) [reply containers/ddev-webserver/ddev-webserver-base-files/usr/local/bin/enable_xhprof:10]: Done, thanks.

File: pkg/ddevapp/config.go [hunk 8]
```diff
@@ -979,6 +999,13 @@ func (app *DdevApp) RenderComposeYAML() (string, error) {
 		// Default max time we wait for containers to be healthy
 		DefaultContainerTimeout: app.DefaultContainerTimeout,
 		StartScriptTimeout:      app.GetStartScriptTimeout(),
+		XHGuiHTTPPort:           app.GetXHGuiHTTPPort(),
+		XHGuiHTTPSPort:          app.GetXHGuiHTTPSPort(),
+		XHGuiPort:               GetInternalPort(app, "xhgui"),
+		HostXHGuiPort:           app.HostXHGuiPort,
+		XhguiImage:              versionconstants.XhguiImage,
+		XHProfMode:              app.XHProfMode,
```
Comment (stasadev) [inline pkg/ddevapp/config.go:1007]: To fix old xhprof:

```suggestion
		XHProfMode:              app.GetXHProfMode(),
```

File: pkg/versionconstants/versionconstants.go [hunk 9]
```diff
@@ -34,6 +34,8 @@ var SSHAuthTag = "v1.24.3"
 // BusyboxImage is used a couple of places for a quick-pull
 var BusyboxImage = "busybox:stable"
 
+var XhguiImage = "ddev/ddev-xhgui:20250307_rfay_xhgui"
```
[the remark this answers is not in the document]
Comment (rfay) [reply pkg/versionconstants/versionconstants.go:37]: I really have wanted to leave the two-variable technique behind. Hate it!

File: cmd/ddev/cmd/xhgui.go [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+// DdevXHGuiCmd represents the xhgui command.
+var DdevXHGuiCmd = &cobra.Command{
+	ValidArgsFunction: ddevapp.GetProjectNamesFunc("active", 1),
+	Use:               "xhgui [on|off|launch|status]",
+	Short:             "Starts or checks status of XHGui performance monitoring and UI",
+	Example: `ddev xhgui
+ddev xhgui launch
+ddev xhgui on
+ddev xhgui off
+ddev xhgui status
```
Comment (rpkoller) [inline cmd/ddev/cmd/xhgui.go:22]: but there is still something in the user experience that lets me stumble. in particular there are that many contexts ddev xhgui is working in (mainly launching xhgui but also providing instructions if not started yet or if the global config is missing the setting) ... one thing i i guess MIGHT help in some way might be adding the commands including their descriptions to the help text. .. at the moment you only have examples without any description and flags but not any commands. i would ask myself what does ddev xhgui actually stand for? launch is clear, same as on off and status. but ddev xhgui is ambiguous and confusing. it always makes me think (and i am drawn to only use explicit subcommands like `ddev xhgui on`. not sure it is just my impression with the interaction with the available set of commands.

File: pkg/versionconstants/versionconstants.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (stasadev) [reply pkg/versionconstants/versionconstants.go]: It's easier for me to see which tags I need to update (looking at the whole variable) when working on a new release.

File: cmd/ddev/cmd/xhgui.go (hunk 10 above)
Comment (stasadev) [reply cmd/ddev/cmd/xhgui.go]: `ddev xhgui` is the same as `ddev xhgui launch` - it works just like it worked in the add-on, so there is no need to remember about launch.

(I pushed a new commit that changed the formatting on the affected lines here, so if you want to change something, please make a new comment.)

**Comment by rfay** [comment]
I added a sleep to the TestCmdXHGui

Improved the `ddev xhgui -h`

**Review by tyler36 — APPROVED** [review]
Tested on `ddev version v1.24.3-95-g915cd248b`.
I didn

- ✅ `ddev xhgui` asks to add global configuration

    - Manually remove `xhprof-mode` from global config

    ```shell
    $ ddev xhgui
    XHProf Mode is set to 'prepend', can't use 'ddev xhgui'.
    Use the command below to enable it:
    ddev config global --xhprof-mode=xhgui && ddev restart
    ```

- ✅ `ddev config global --xhprof-mode=xhgui` updates config

    ```shell
    $ ddev config global --xhprof-mode=xhgui && ddev restart
    ...
    Your project can be reached at https://lara11-base-demo.ddev.site
    ```

- ✅ `ddev xhgui` on starts debugging, confirmed with `ddev xhgui launch`

    ```shell
    $ ddev xhgui                                            
     Container ddev-lara11-base-demo-xhgui  Created 
     Container ddev-lara11-base-demo-xhgui  Started 
    Started optional compose profiles 'xhgui' 
    Enabled XHProf for XHGui. 
    ```

- ✅ `ddev xhgui status` correctly reports on

    ```shell
    $ ddev xhgui status
    XHProf is enabled and capturing performance information. 
    The XHGui service is running and you can access it at https://lara11-base-demo.ddev.site:8142 
    ```

- ✅ `ddev xhgui off` and am no longer able to access website via ddev xhgui lanuch

    ```shell
    $ ddev xhgui off 
    $ ddev xhgui status
    XHProf is disabled. 
    XHGui is disabled. 
    ```

- ✅ delete project database, changed to Postgres 16 and correctly ran ddev xhgui launch

    ```shell
    $ ddev delete -Oy
    $ ddev config --project-type=php --database=postgres:16
    $ ddev xhgui
    ```

- ✅ `ddev xhgui -h` and `ddev xhgui --help` both correctly display help messgae

    ```shell
    $ ddev xhgui
    ddev xhgui -h    
      Starts or checks status of XHGui/XHProf performance monitoring and UI.
      You can use 'ddev xhgui on' to enable, 'ddev xhgui' or 'ddev xhgui launch' to view the UI.
      'ddev xhgui status' shows the status.
    ...
    ```

- ✅ `ddev xhgui -j` does nothing (previously opened browser)

**Review by stasadev — CHANGES_REQUESTED** [review]
- [x] Install ddev/ddev-xhgui and then ddev start to see removal
- [x] Validate that xhprof still works with xhprof-mode=prepend
- [x] tested it with `mysql:8.4`
- [x] tested it without `db` container, seeing an expected error on `ddev xhgui`

My nitpicks:

- We should have a named volume for xhgui https://github.com/perftools/xhgui/issues/505

- The xhgui image is pulled up during the first `ddev start`, which may slow down some CI tests, but on the other hand, people may see it and wonder "What is this new image? What does it do?" and start using it, which can give us more feedback on `ddev xhgui`

- The port configuration can only be done with manual editing for `.ddev/config.yaml`, I don't expect people will use this often, so not asking for new flags in `ddev config`

- Testing without router worked only when I added `host_xhgui_port: 8142`, without it opens the main website (so the host port is not assigned automatically)

- Not related to this PR, but without router `ddev mailpit` opens `http://127.0.0.1:8025`, and not the actual port from `ddev describe`

**Review by stasadev — APPROVED** [review]
Tested the named volume, it works for me.
